flang/lib/Lower/VectorSubscripts.cpp

```diff
@@ -0,0 +1,428 @@
+  namedEntityToExpr(const Fortran::evaluate::NamedEntity &namedEntity) {
+    return Fortran::evaluate::AsGenericExpr(namedEntityToDataRef(namedEntity))
+        .value();
+  }
+
+  Fortran::lower::AbstractConverter &converter;
+  Fortran::lower::StatementContext &stmtCtx;
+  mlir::Location loc;
+  /// Elements of VectorSubscriptBox being built.
+  fir::ExtendedValue loweredBase;
+  llvm::SmallVector<LoweredSubscript, 16> loweredSubscripts;
+  llvm::SmallVector<mlir::Value> componentPath;
+  MaybeSubstring substringBounds;
+  mlir::Type elementType;
+};
+} // namespace
+
+Fortran::lower::VectorSubscriptBox Fortran::lower::genVectorSubscriptBox(
+    mlir::Location loc, Fortran::lower::AbstractConverter &converter,
+    Fortran::lower::StatementContext &stmtCtx,
+    const Fortran::evaluate::Expr<Fortran::evaluate::SomeType> &expr) {
+  return VectorSubscriptBoxBuilder(loc, converter, stmtCtx).gen(expr);
+}
+
+template <typename LoopType, typename Generator>
+mlir::Value Fortran::lower::VectorSubscriptBox::loopOverElementsBase(
+    fir::FirOpBuilder &builder, mlir::Location loc,
+    const Generator &elementalGenerator, mlir::Value initialCondition) {
+  auto shape = builder.createShape(loc, loweredBase);
+  auto slice = createSlice(builder, loc);
+
+  // Create loop nest for triplets and vector subscripts in column
+  // major order.
+  llvm::SmallVector<mlir::Value> inductionVariables;
+  LoopType outerLoop;
+  for (auto [lb, ub, step] : genLoopBounds(builder, loc)) {
+    LoopType loop;
+    if constexpr (std::is_same_v<LoopType, fir::IterWhileOp>) {
+      loop =
+          builder.create<fir::IterWhileOp>(loc, lb, ub, step, initialCondition);
+      initialCondition = loop.getIterateVar();
+      if (!outerLoop)
+        outerLoop = loop;
+      else
+        builder.create<fir::ResultOp>(loc, loop.getResult(0));
+    } else {
+      loop =
+          builder.create<fir::DoLoopOp>(loc, lb, ub, step, /*unordered=*/false);
+      if (!outerLoop)
+        outerLoop = loop;
+    }
+    builder.setInsertionPointToStart(loop.getBody());
+    inductionVariables.push_back(loop.getInductionVar());
+  }
+  assert(outerLoop && !inductionVariables.empty() &&
+         "at least one loop should be created");
+
+  auto elem = getElementAt(builder, loc, shape, slice, inductionVariables);
+
+  if constexpr (std::is_same_v<LoopType, fir::IterWhileOp>) {
+    auto res = elementalGenerator(elem);
+    builder.create<fir::ResultOp>(loc, res);
+    builder.setInsertionPointAfter(outerLoop);
+    return outerLoop.getResult(0);
+  } else {
+    elementalGenerator(elem);
+    builder.setInsertionPointAfter(outerLoop);
+    return {};
+  }
+}
+
+void Fortran::lower::VectorSubscriptBox::loopOverElements(
+    fir::FirOpBuilder &builder, mlir::Location loc,
+    const ElementalGenerator &elementalGenerator) {
+  mlir::Value initialCondition;
+  loopOverElementsBase<fir::DoLoopOp, ElementalGenerator>(
+      builder, loc, elementalGenerator, initialCondition);
```

Comment:
When I run `check-flang` after doing a build, I get a warning here.  And since I have warnings treated as errors, `check-flang` fails.  Here's the relevant excerpt from the log file:
```
[71/84] Building CXX object lib/Lower/CMakeFiles/obj.FortranLower.dir/VectorSubscripts.cpp.o
FAILED: lib/Lower/CMakeFiles/obj.FortranLower.dir/VectorSubscripts.cpp.o 
/usr/bin/c++ -DGTEST_HAS_RTTI=0 -D_DEBUG -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/include -I../include -Iinclude -isystem /local/home/psteinfeld/cg/clean/install/include -isystem /local/home/psteinfeld/cg/b1/llvm/include -isystem /local/home/psteinfeld/cg/b1/build/include -fPIC -fvisibility-inlines-hidden -Werror=date-time -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wno-class-memaccess -Wno-noexcept-type -Wdelete-non-virtual-dtor -Wno-comment -Wmisleading-indentation -fdiagnostics-color -ffunction-sections -fdata-sections -Werror -Wno-deprecated-copy -fno-strict-aliasing -fno-semantic-interposition -O3 -DNDEBUG   -D_GNU_SOURCE -D_DEBUG -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS  -fno-exceptions -fno-rtti -UNDEBUG -std=c++17 -MD -MT lib/Lower/CMakeFiles/obj.FortranLower.dir/VectorSubscripts.cpp.o -MF lib/Lower/CMakeFiles/obj.FortranLower.dir/VectorSubscripts.cpp.o.d -o lib/Lower/CMakeFiles/obj.FortranLower.dir/VectorSubscripts.cpp.o -c ../lib/Lower/VectorSubscripts.cpp
../lib/Lower/VectorSubscripts.cpp: In instantiation of â€˜mlir::Value Fortran::lower::VectorSubscriptBox::loopOverElementsBase(fir::FirOpBuilder&, mlir::Location, const Generator&, mlir::Value) [with LoopType = fir::DoLoopOp; Generator = std::function<void(const fir::ExtendedValue&)>]â€™:
../lib/Lower/VectorSubscripts.cpp:308:57:   required from here
../lib/Lower/VectorSubscripts.cpp:259:54: error: parameter â€˜initialConditionâ€™ set but not used [-Werror=unused-but-set-parameter]
     const Generator &elementalGenerator, mlir::Value initialCondition) {
                                          ~~~~~~~~~~~~^~~~~~~~~~~~~~~~
cc1plus: error: unrecognized command line option â€˜-Wno-deprecated-copyâ€™ [-Werror]
cc1plus: all warnings being treated as errors
```

Reply:
I added a [[maybe_unused]] tag to the argument (line 260). The version of clang I used for build (llvm8) did not warn. I think it's a bit annoying that modern clang warns here when the parameter is used inside if constexpr... Anyway, should be fixed.